core/src/main/scala/org/apache/spark/util/DependencyUtils.scala

```diff
@@ -15,22 +15,112 @@
  * limitations under the License.
  */
 
-package org.apache.spark.deploy
+package org.apache.spark.util
 
 import java.io.File
-import java.net.URI
+import java.net.{URI, URISyntaxException}
 
 import org.apache.commons.lang3.StringUtils
 import org.apache.hadoop.conf.Configuration
 import org.apache.hadoop.fs.{FileSystem, Path}
 
 import org.apache.spark.{SecurityManager, SparkConf, SparkException}
+import org.apache.spark.deploy.SparkSubmitUtils
 import org.apache.spark.internal.Logging
-import org.apache.spark.util.{MutableURLClassLoader, Utils}
 
-private[deploy] object DependencyUtils extends Logging {
+private[spark] object DependencyUtils extends Logging {
+
+  def getIvyProperties(): Seq[String] = {
+    Seq(
+      "spark.jars.excludes",
+      "spark.jars.packages",
+      "spark.jars.repositories",
+      "spark.jars.ivy",
+      "spark.jars.ivySettings"
+    ).map(sys.props.get(_).orNull)
+  }
+
+  private def parseQueryParams(uriQuery: String): (Boolean, String) = {
+    if (uriQuery == null) {
+      (false, "")
+    } else {
+      val mapTokens = uriQuery.split("&").map(_.split("="))
+      if (mapTokens.exists(_.length != 2)) {
+        throw new URISyntaxException(uriQuery, s"Invalid query string: $uriQuery")
+      }
+      val groupedParams = mapTokens.map(kv => (kv(0), kv(1))).groupBy(_._1)
+      // Parse transitive parameters (e.g., transitive=true) in an ivy URL, default value is false
+      var transitive = false
```

Comment:
To be a bit more functional / Scala-idiomatic, I think we can do something like `groupedParams.get("transitive").takeRight(1).headOption.getOrElse(false)` instead of the `foreach` call

Reply:
> To be a bit more functional / Scala-idiomatic, I think we can do something like `groupedParams.get("transitive").takeRight(1).headOption.getOrElse(false)` instead of the `foreach` call

In this way will lose warning message?

Reply:
Maybe:
```
val transitiveParams = groupedParams.get("transitive")
if (transitiveParams.map(_.size).getOrElse(0) > 1) {
  // log warning
}
val transitive = transitiveParams.flatMap(_.takeRight(1).headOption).getOrElse(false)
```

Reply:
Should do like this:
```
    val transitiveParams = groupedParams.get("transitive")
      if (transitiveParams.map(_.size).getOrElse(0) > 1) {
        logWarning("It's best to specify `transitive` parameter in ivy URL query only once." +
          " If there are multiple `transitive` parameter, we will select the last one")
      }
      val transitive = transitiveParams.flatMap(_.takeRight(1).map { case (_, value) =>
        value match {
          case "true" => true
          case _ => false
        }
      }.headOption).getOrElse(false)
```

is equal with current logic.

Reply:
Whoops, forgot about having to unpack the value. To be a bit more concise you can also do:
```
val transitive = transitiveParams.flatMap(_.takeRight(1).map(_._2 == "true").headOption).getOrElse(false)
```
But I think either way looks good!

Reply:
Yea, I have too few ideas on how to make code more concise.
Simpler is better, Updated!